kani-driver/src/cbmc_output_parser.rs

```diff
@@ -0,0 +1,991 @@
+    Unreachable,
+}
+
+impl std::fmt::Display for CheckStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let check_str = match self {
+            CheckStatus::Success => style("SUCCESS").green(),
+            CheckStatus::Failure => style("FAILURE").red(),
+            CheckStatus::Unreachable => style("UNREACHABLE").yellow(),
+            CheckStatus::Undetermined => style("UNDETERMINED").yellow(),
+        };
+        write! {f, "{}", check_str}
+    }
+}
+
+#[derive(PartialEq)]
+enum Action {
+    ClearInput,
+    ProcessItem,
+}
+
+/// A parser for CBMC output, whose state is determined by:
+///  1. The input accumulator, required to process items on the fly.
+///  2. The buffer, which is accessed to retrieve more lines.
+///
+/// CBMC's JSON output is defined as a JSON array which contains:
+///  1. One program at the beginning (i.e., a message with CBMC's version).
+///  2. Messages, which can appear anywhere, and are either status messages or error messages.
+///  3. Verification results, another JSON array with all individual checks.
+///  4. Prover status, at the end. Because the verification results depends on
+///     our postprocessing, this is not used.
+///
+/// The parser reads the output line by line. A line may trigger one action, and
+/// the action may return a parsed item.
+struct Parser<'a, 'b> {
+    pub input_so_far: String,
+    pub buffer: &'a mut BufReader<&'b mut ChildStdout>,
+}
+
+impl<'a, 'b> Parser<'a, 'b> {
+    pub fn new(buffer: &'a mut BufReader<&'b mut ChildStdout>) -> Self {
+        Parser { input_so_far: String::new(), buffer }
+    }
+
+    /// Triggers an action based on the input:
+    ///  * Square brackets ('[' and ']') will trigger the `ClearInput` action
+    ///    because we assume parsing is done on a JSON array.
+    ///  * Curly closing bracket ('}') preceded by two spaces will trigger the
+    ///    `ProcessItem` action.
+    fn triggers_action(&self, input: String) -> Option<Action> {
+        if input.starts_with('[') || input.starts_with(']') {
+            return Some(Action::ClearInput);
+        }
+        if input.starts_with("  }") {
+            return Some(Action::ProcessItem);
+        }
+        None
+    }
+
+    /// Clears the input accumulated so far.
+    fn clear_input(&mut self) {
+        self.input_so_far = String::new();
+    }
+
+    /// Performs an action. In both cases, the input is cleared.
+    fn do_action(&mut self, action: Action) -> Option<ParserItem> {
+        match action {
+            Action::ClearInput => {
+                self.clear_input();
+                None
+            }
+            Action::ProcessItem => {
+                let item = self.parse_item();
+                self.clear_input();
+                Some(item)
+            }
+        }
+    }
+
+    // Adds a string to the input accumulated so far
+    fn add_to_input(&mut self, input: String) {
+        self.input_so_far.push_str(input.as_str());
```

Comment:
can't you change `x.as_str()` to `&x` if `x: String`?

Reply:
Yes, thanks! Replaced almost all of them.

Reply:
AFAIK, `x.as_str()` is the same as `&x`: https://users.rust-lang.org/t/converting-between-str-and-string/53595

Reply:
Yes, but `&x` is easier to read and less noisy IMO

Reply:
Yeah, I wasn't aware of this implicit conversion earlier. It's explained in detail here: https://doc.rust-lang.org/std/string/struct.String.html#deref